packages/instrumentation-llamaindex/src/custom-llm-instrumentation.ts

```diff
@@ -0,0 +1,175 @@
+                  result.message.content,
+                );
+              }
+              span.setStatus({ code: SpanStatusCode.OK });
+              span.end();
+              resolve(result);
+            });
+          })
+          .catch((error: Error) => {
+            return new Promise((_, reject) => {
+              span.setStatus({
+                code: SpanStatusCode.ERROR,
+                message: error.message,
+              });
+              span.end();
+              reject(error);
+            });
+          });
+        return context.bind(execContext, wrappedPromise as any);
+      };
+    };
+  }
+
+  chatWrapper({ className }: { className: string }) {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const plugin = this;
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    return (original: Function) => {
+      return function method(this: llamaindex.LLM, ...args: unknown[]) {
+        const messages = args[0] as llamaindex.ChatMessage[];
+
+        const span = plugin.tracer.startSpan(
+          `${lodash.snakeCase(className)}.chat`,
+        );
+
+        span.setAttribute(SpanAttributes.LLM_VENDOR, "llamaindex");
+        span.setAttribute(
+          SpanAttributes.LLM_REQUEST_MODEL,
+          this.metadata.model,
+        );
+        span.setAttribute(SpanAttributes.LLM_REQUEST_TYPE, "chat");
+        span.setAttribute(SpanAttributes.LLM_TOP_P, this.metadata.topP);
+        if (shouldSendPrompts(plugin.config)) {
+          for (const messageIdx in messages) {
+            span.setAttribute(
+              `${SpanAttributes.LLM_PROMPTS}.${messageIdx}.content`,
+              messages[messageIdx].content,
+            );
+            span.setAttribute(
+              `${SpanAttributes.LLM_PROMPTS}.${messageIdx}.role`,
+              messages[messageIdx].role,
+            );
+          }
+        }
+
+        const execContext = trace.setSpan(context.active(), span);
+        const execPromise = safeExecuteInTheMiddle(
+          () => {
+            return context.with(execContext, () => {
+              return original.apply(this, args);
+            });
+          },
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          (error) => {},
+        );
+        const wrappedPromise = execPromise
+          .then((result: any) => {
+            return new Promise((resolve) => {
+              span.setAttribute(
+                SpanAttributes.LLM_RESPONSE_MODEL,
+                this.metadata.model,
+              );
+              if (shouldSendPrompts(plugin.config)) {
+                span.setAttribute(
+                  `${SpanAttributes.LLM_COMPLETIONS}.0.role`,
+                  result.message.role,
+                );
+                span.setAttribute(
+                  `${SpanAttributes.LLM_COMPLETIONS}.0.content`,
+                  result.message.content,
+                );
```

Comment:
2 points here:
1. Are you sure `role` exist here?
2. There can't be more than one completion?

Reply:
> 1. Are you sure role exist here?

Yes, it's a non optional field of ChatMessage Interface.

```typescript
interface ChatMessage {
    content: any;
    role: MessageType;
}
```

> 2. There can't be more than one completion?

Just checked return type of chat, it could also return AsyncGenerator. I'll update chat and completion wrapper to handle streaming response.

Reply:
@galkleinman I've pushed changes to add support for capturing streaming response.